browser/registry.go

```diff
@@ -148,38 +154,136 @@ func (r *remoteRegistry) isRemoteBrowser() (string, bool) {
 	return wsURL, true
 }
 
-// browserRegistry stores browser instances indexed per
-// iteration as identified by VUID-scenario-iterationID.
+// browserRegistry stores a single VU browser instances
+// indexed per iteration.
 type browserRegistry struct {
+	vu k6modules.VU
+
 	mu sync.RWMutex
-	m  map[string]api.Browser
+	m  map[int64]api.Browser
+
+	buildFn browserBuildFunc
+}
+
+type browserBuildFunc func(ctx context.Context) (api.Browser, error)
+
+func newBrowserRegistry(vu k6modules.VU, remote *remoteRegistry, pids *pidRegistry) *browserRegistry {
+	bt := chromium.NewBrowserType(vu)
+	builder := func(ctx context.Context) (api.Browser, error) {
+		var (
+			err                    error
+			b                      api.Browser
+			wsURL, isRemoteBrowser = remote.isRemoteBrowser()
+		)
+
+		if isRemoteBrowser {
+			b, err = bt.Connect(ctx, wsURL)
+			if err != nil {
+				return nil, err //nolint:wrapcheck
+			}
+		} else {
+			var pid int
+			b, pid, err = bt.Launch(ctx)
+			if err != nil {
+				return nil, err //nolint:wrapcheck
+			}
+			pids.registerPid(pid)
+		}
+
+		return b, nil
+	}
+
+	r := &browserRegistry{
+		vu:      vu,
+		m:       make(map[int64]api.Browser),
+		buildFn: builder,
+	}
+
+	iterSubID, eventsCh := vu.Events().Local.Subscribe(
+		k6event.IterStart,
+		k6event.IterEnd,
+	)
+	unsubscribe := func() {
+		vu.Events().Local.Unsubscribe(iterSubID)
+	}
+	go r.handleIterEvents(eventsCh, unsubscribe)
+
+	return r
 }
 
-func newBrowserRegistry() *browserRegistry {
-	return &browserRegistry{
-		m: make(map[string]api.Browser),
+func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsubscribeFn func()) {
+	var (
+		ok    bool
+		data  k6event.IterData
+		ctx   = context.Background()
+		vuCtx = k6ext.WithVU(r.vu.Context(), r.vu)
+	)
+
+	for e := range eventsCh {
+		// If browser module is imported in the test, NewModuleInstance will be called for
+		// every VU. Because on VU init stage we can not distinguish to which scenario it
+		// belongs or access its options (because state is nil), we have to always subscribe
+		// to each VU iter events, including VUs that do not make use of the browser in their
+		// iterations.
+		// Therefore, if we get an event that does not correspond to a browser iteration, then
+		// unsubscribe for the VU events and exit the loop in order to reduce unuseful overhead.
+		if !isBrowserIter(r.vu) {
```

Comment:
Thanks for the suggestion! I could see some good points for this option so I've been experimenting with it. Unfortunately the VU state is no set in the `Init` event, so we can not access the options to determine if browser type is set. We should take this into account for future discussions with k6-core when considering refactor options for the events handling, but as of now I think we should stick to the current implementation.

<details>
<summary>See the diff with debug</summary>

```diff
diff --git a/browser/registry.go b/browser/registry.go
index ebebbf7..8671d98 100644
--- a/browser/registry.go
+++ b/browser/registry.go
@@ -204,24 +204,43 @@ func newBrowserRegistry(vu k6modules.VU, remote *remoteRegistry, pids *pidRegist
        }
 
        exitSubID, exitCh := vu.Events().Global.Subscribe(
+               k6event.Init,
                k6event.Exit,
        )
-       go r.handleExitEvent(exitCh)
-
-       iterSubID, eventsCh := vu.Events().Local.Subscribe(
-               k6event.IterStart,
-               k6event.IterEnd,
-       )
        unsubscribe := func() {
-               vu.Events().Local.Unsubscribe(iterSubID)
-               vu.Events().Global.Unsubscribe(exitSubID)
+               r.vu.Events().Global.Unsubscribe(exitSubID)
        }
-       go r.handleIterEvents(eventsCh, unsubscribe)
+       go r.handleGlobalEvents(exitCh, unsubscribe)
 
        return r
 }
 
-func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsubscribeFn func()) {
+func (r *browserRegistry) handleGlobalEvents(eventsCh <-chan *k6event.Event, unsubscribeFn func()) {
+       for e := range eventsCh {
+               switch e.Type { //nolint:exhaustive
+               case k6event.Init:
+                       fmt.Println("RECEIVED INIT")
+                       if !isBrowserVU(r.vu) {
+                               e.Done()
+                               unsubscribeFn()
+                               return
+                       }
+
+                       fmt.Println("SUBSCRIBING TO ITER EVENTS")
+                       _, eventsCh := r.vu.Events().Local.Subscribe(
+                               k6event.IterStart,
+                               k6event.IterEnd,
+                       )
+                       r.handleIterEvents(eventsCh)
+                       e.Done()
+               case k6event.Exit:
+                       r.clear()
+                       e.Done()
+               }
+       }
+}
+
+func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event) {
        var (
                ok    bool
                data  k6event.IterData
@@ -230,20 +249,6 @@ func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsub
        )
 
        for e := range eventsCh {
-               // If browser module is imported in the test, NewModuleInstance will be called for
-               // every VU. Because on VU init stage we can not distinguish to which scenario it
-               // belongs or access its options (because state is nil), we have to always subscribe
-               // to each VU iter events, including VUs that do not make use of the browser in their
-               // iterations.
-               // Therefore, if we get an event that does not correspond to a browser iteration, then
-               // unsubscribe for the VU events and exit the loop in order to reduce unuseful overhead.
-               if !isBrowserIter(r.vu) {
-                       unsubscribeFn()
-                       r.stop()
-                       e.Done()
-                       return
-               }
-
                if data, ok = e.Data.(k6event.IterData); !ok {
                        e.Done()
                        k6ext.Abort(vuCtx, "unexpected iteration event data format: %v", e.Data)
@@ -255,6 +260,7 @@ func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsub
 
                switch e.Type { //nolint:exhaustive
                case k6event.IterStart:
+                       fmt.Println("ITER START")
                        b, err := r.buildFn(ctx)
                        if err != nil {
                                e.Done()
@@ -266,6 +272,7 @@ func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsub
                        }
                        r.setBrowser(data.Iteration, b)
                case k6event.IterEnd:
+                       fmt.Println("ITER END")
                        r.deleteBrowser(data.Iteration)
                default:
                        r.vu.State().Logger.Warnf("received unexpected event type: %v", e.Type)
@@ -275,15 +282,6 @@ func (r *browserRegistry) handleIterEvents(eventsCh <-chan *k6event.Event, unsub
        }
 }

-func (r *browserRegistry) handleExitEvent(exitCh <-chan *k6event.Event) {
-       e, ok := <-exitCh
-       if !ok {
-               return
-       }
-       defer e.Done()
-       r.clear()
-}
-
 func (r *browserRegistry) setBrowser(id int64, b api.Browser) {
        r.mu.Lock()
        defer r.mu.Unlock()
@@ -322,12 +320,9 @@ func (r *browserRegistry) clear() {
        }
 }
 
-func (r *browserRegistry) stop() {
-       r.stopped.Store(true)
-}
-
-func isBrowserIter(vu k6modules.VU) bool {
+func isBrowserVU(vu k6modules.VU) bool {
        opts := k6ext.GetScenarioOpts(vu.Context(), vu)
+       fmt.Printf("SCENARIO OPTS: %v\n", opts)
        _, ok := opts["type"] // Check if browser type option is set
        return ok
 }
diff --git a/k6ext/context.go b/k6ext/context.go
index 8f763bc..4a996eb 100644
--- a/k6ext/context.go
+++ b/k6ext/context.go
@@ -2,6 +2,7 @@ package k6ext
 
 import (
        "context"
+       "fmt"
 
        k6modules "go.k6.io/k6/js/modules"
        k6lib "go.k6.io/k6/lib"
@@ -67,9 +68,11 @@ func GetScenarioName(ctx context.Context) string {
 func GetScenarioOpts(ctx context.Context, vu k6modules.VU) map[string]any {
        scenario := GetScenarioName(ctx)
        if scenario == "" {
+               fmt.Println("RETURNING NIL")
                return nil
        }
        if so := vu.State().Options.Scenarios[scenario].GetScenarioOptions(); so != nil {
+               fmt.Println("RETURNING OPTS")
                return so.Browser
        }
        return nil
```

</details>

What do you think?

Reply:
Thanks for taking the time to investigate the suggestion! Yep, if it's not possible to retrieve the `state` on `init`, then the current solution is good, and we can work with k6 core to help us optimise later 👍